checks/src/main/kotlin/com/faithlife/lint/ForEachFunctionDetector.kt

```diff
@@ -10,23 +10,48 @@ import com.android.tools.lint.detector.api.Scope
 import com.android.tools.lint.detector.api.Severity
 import com.android.tools.lint.detector.api.SourceCodeScanner
 import com.intellij.psi.PsiMethod
+import org.jetbrains.kotlin.psi.KtLambdaExpression
 import org.jetbrains.uast.UCallExpression
 import org.jetbrains.uast.kotlin.KotlinUSafeQualifiedExpression
 
 class ForEachFunctionDetector : Detector(), SourceCodeScanner {
-    override fun getApplicableMethodNames(): List<String> = listOf("forEach", "forEachIndexed")
+    override fun getApplicableMethodNames(): List<String> =
+        listOf(FOR_EACH_NAME, FOR_EACH_INDEXED_NAME)
 
     override fun visitMethodCall(context: JavaContext, node: UCallExpression, method: PsiMethod) {
         if (node.uastParent is KotlinUSafeQualifiedExpression) return
 
+        val receiver = node.receiver
         val receiverTypeClass = context.evaluator.getTypeClass(node.receiverType)
 
-        if (receiverTypeClass != null && context.evaluator.inheritsFrom(receiverTypeClass, "java.lang.Iterable", false)) {
+        if (receiver != null && receiverTypeClass != null && context.evaluator.inheritsFrom(receiverTypeClass, "java.lang.Iterable", false)) {
+            val lambda = node.valueArguments.first().sourcePsi as KtLambdaExpression
```

Comment:
It's generally better to deal with UAST types over PSI. I suspect being careful to do so might help us avoid issues with the k2 toolchain when it's ready (which doesn't use PSI internally anymore).

`ULambdaExpression` or `KotlinULambdaExpression` would do the job.

Reply:
We probably want to also either: a) avoid a hard cast that assumes kotlin source b) bail early on this entire lint check if the source isn't kotlin. One could conceivably write a forEach function on a type that extends `Iterable` in Java.

Since assumptions are being made about the parameters of forEach and forEachIndexed, the second seems a little safer.

Reply:
I couldn't figure out how to get the correct range with the UAST lambda types. It makes sense that UAST is preferred but PSI seems baked into a lot of the UAST elements and `visitMethodCall` even has a PSI parameter.

Reply:
> One could conceivably write a forEach function on a type that extends Iterable in Java.

That's also possible in Kotlin so I added a check on the package of the method.

Reply:
> but PSI seems baked into a lot of the UAST elements and visitMethodCall even has a PSI parameter.

There's history there. Sometimes using PSI is unavoidable, especially w.r.t. method representation. https://googlesamples.github.io/android-custom-lint-rules/api-guide.html#writingalintcheck:basics/analyzingkotlinandjavacode/psi